### PR TITLE
fix(VersionMonitor): improper version update to older version

### PR DESCRIPTION
**Pull request by kostasrim**
Resolves: #1449

### CODE REVIEW

**Review by adiholden — APPROVED** [review]

File: src/server/dfly_main.cc [hunk 1]
```diff
@@ -206,7 +206,7 @@ void VersionMonitor::RunTask(SSL_CTX* ssl_ctx) {
         GetRemoteVersion(my_pb, ssl_ctx, host_name, port, resource, version_header);
     if (remote_version) {
       const std::string rv = remote_version.value();
-      if (rv != current_version) {
```
Comment (chakaz) [inline src/server/dfly_main.cc:209]: I'm not sure if this is the best solution. We're using `X.Y.Z` format, and it isn't string sortable. Think of the case of current == `1.6.1` vs rv == `1.12.1`.
I guess we'll need to `absl::StrSplit()` it into 3 parts and compare each?

Comment (kostasrim) [reply src/server/dfly_main.cc:209]: Really good catch. I am not sure how I missed this but I am very happy that you spotted this.

File: src/server/dfly_main.cc [hunk 2]
```diff
@@ -161,9 +163,28 @@ struct VersionMonitor {
...
  private:
+  bool IsVersionOutdated(std::string_view remote, std::string_view current) const;
   void RunTask(SSL_CTX* ssl_ctx);
 };
 
+bool VersionMonitor::IsVersionOutdated(const std::string_view remote,
+                                       const std::string_view current) const {
+  const std::vector<absl::string_view> remote_xyz = absl::StrSplit(remote, ".");
+  const std::vector<absl::string_view> current_xyz = absl::StrSplit(current, ".");
+  CHECK_EQ(remote_xyz.size(), current_xyz.size());
```
Comment (chakaz) [inline src/server/dfly_main.cc:173]: I wouldn't crash if the remote responds with a weird version. A bad push on our side can effectively crash all Dragonfly instances in the world :worried: 
I would perhaps `LOG` some warning and return that the version is not outdated.

File: src/server/dfly_main.cc [hunk 3]
```diff
@@ -161,9 +162,28 @@ struct VersionMonitor {
...
+  for (size_t i = 0; i < remote_xyz.size(); ++i) {
+    size_t remote_x = 0;
+    size_t current_x = 0;
+    CHECK_EQ(absl::SimpleAtoi(remote, &remote_x), true);
+    CHECK_EQ(absl::SimpleAtoi(current, &current_x), true);
```
Comment (chakaz) [inline src/server/dfly_main.cc:178]: You could just `CHECK()` for boolean expressions, no need to compare to `true`.
However, again, I would not crash if the `remote` part is not a legal number.

File: src/server/dfly_main.cc [hunk 4]
```diff
@@ -161,9 +162,33 @@ struct VersionMonitor {
...
+  bool IsVersionOutdated(std::string_view remote, std::string_view current) const;
   void RunTask(SSL_CTX* ssl_ctx);
 };
 
+bool VersionMonitor::IsVersionOutdated(const std::string_view remote,
+                                       const std::string_view current) const {
+  const std::vector<absl::string_view> remote_xyz = absl::StrSplit(remote, ".");
+  const std::vector<absl::string_view> current_xyz = absl::StrSplit(current, ".");
+  if (remote_xyz.size() != current_xyz.size()) {
+    LOG(WARNING) << "Remote and current version fields must be of equal size";
```
Comment (chakaz) [inline src/server/dfly_main.cc:174]: Nit: I'd rephrase this message to something that can provide more context to users, who will see this printed out of nowhere.
Maybe: `"Can't compare Dragonfly version " << current << " to latest version " << remote;` ?

File: src/server/dfly_main.cc [hunk 5]
```diff
@@ -161,9 +162,33 @@ struct VersionMonitor {
...
+    return false;
+  }
+  for (size_t i = 0; i < remote_xyz.size(); ++i) {
+    size_t remote_x = 0;
+    size_t current_x = 0;
+    if (!absl::SimpleAtoi(remote, &remote_x) || !absl::SimpleAtoi(current, &current_x)) {
```
Comment (chakaz) [inline src/server/dfly_main.cc:180]: Don't you want to parse `remote_xyz[i]` instead of just `remote`?
(same for `current`)

File: src/server/dfly_main.cc [hunk 6]
```diff
@@ -161,9 +162,33 @@ struct VersionMonitor {
...
+      LOG(WARNING) << "Conversion error on version string_view";
```
Comment (chakaz) [inline src/server/dfly_main.cc:181]: Same here, especially with "string_view" which will confuse readers I think :)
Maybe `Can't parse " << ... << " part of version " << remote << " as a number";` ?

**Review by chakaz — APPROVED** [review]
LG, thanks for bearing with me! :)
